File: Nololiyt.IndexedDbSharp.Raw/Nololiyt.IndexedDbSharp.Raw/Typescripts/Wrapping/WrappedIdbTransaction.ts
import { DotNetCallbackObject } from "../Entities/DotNetCallbackObject.js";
import { WrappedDomException } from "./WrappedDomException.js";
import { WrappedEvent } from "./WrappedEvent.js";
import { WrappedIdbDatabase } from "./WrappedIDbDatabase.js";
import { WrappedIdbObjectStore } from "./WrappedIdbObjectStore.js";

export class WrappedIdbTransaction
{
    private wrapped;
    constructor(wrapped: IDBTransaction)
    {
        this.wrapped = wrapped;
    }

    abort(): void
    {
        this.wrapped.abort();
    }

    commit(): void
    {
        this.wrapped.commit();
    }

    db(): WrappedIdbDatabase
    {
        const result = this.wrapped.db;
        return new WrappedIdbDatabase(result);
    }

    durability(): IDBTransactionDurability
    {
        return this.wrapped.durability;
    }

    error(): WrappedDomException | null
    {
        const result = this.wrapped.error;
        if (result)
            return new WrappedDomException(result);
        return null;
    }

    mode(): IDBTransactionMode
    {
        return this.wrapped.mode;
    }

    objectStore(name: string): WrappedIdbObjectStore
    {
        const result = this.wrapped.objectStore(name);
        return new WrappedIdbObjectStore(result);
    }

    objectStoreNames(): string[]
    {
        const names = this.wrapped.objectStoreNames;
        const result = [];
        for (const name of names)
            result.push(name);
        return result;
    }

    setOnAbort(callbackObject: DotNetCallbackObject | null): void
    {
        if (callbackObject)
            this.wrapped.onabort = function (this: IDBTransaction, ev: Event)
            {
                const wrappedThis = new WrappedIdbTransaction(this);
                const wrappedEv = new WrappedEvent(ev);
                callbackObject.InvokeCallback(wrappedThis, wrappedEv);
            };
        else
            this.wrapped.onabort = null;
    }

    setOnComplete(callbackObject: DotNetCallbackObject | null): void
    {
        if (callbackObject)
            this.wrapped.oncomplete = function (this: IDBTransaction, ev: Event)
            {
                const wrappedThis = new WrappedIdbTransaction(this);
                const wrappedEv = new WrappedEvent(ev);
                callbackObject.InvokeCallback(wrappedThis, wrappedEv);
            };
        else
            this.wrapped.oncomplete = null;
    }

    setOnError(callbackObject: DotNetCallbackObject | null): void
    {
        if (callbackObject)
            this.wrapped.onerror = function (this: IDBTransaction, ev: Event)
            {
                const wrappedThis = new WrappedIdbTransaction(this);
                const wrappedEv = new WrappedEvent(ev);
                callbackObject.InvokeCallback(wrappedThis, wrappedEv);
            };
        else
            this.wrapped.onerror = null;
    }
}

File: Nololiyt.IndexedDbSharp.Raw/Nololiyt.IndexedDbSharp.Raw/CSharp/Async/Implementation/Wrapping/WrappedIdbTransaction.cs
using Microsoft.JSInterop;
using Nololiyt.IndexedDbSharp.Raw.CSharp.Async.EventObjects;
using Nololiyt.IndexedDbSharp.Raw.CSharp.Async.Wrapping;
using Nololiyt.IndexedDbSharp.Raw.CSharp.Entities;
using System.Xml.Linq;

namespace Nololiyt.IndexedDbSharp.Raw.CSharp.Async.Implementation.Wrapping
{
    internal sealed class WrappedIdbTransaction :  WrappedWrappedJsObjectBase, IWrappedIdbTransaction
    {
        public WrappedIdbTransaction(IJSObjectReference wrappedObject) : base(wrappedObject)
        {
        }

        public async ValueTask AbortAsync()
        {
            await this.WrappedObject.InvokeVoidAsync("abort");
        }

        public async ValueTask CommitAsync()
        {
            await this.WrappedObject.InvokeVoidAsync("commit");
        }

        public async ValueTask<IWrappedIdbDatabase> GetDbAsync()
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("db");
            return new WrappedIdbDatabase(result);
        }

        public async ValueTask<IdbTransactionDurability> GetDurabilityAsync()
        {
            var result = await this.WrappedObject.InvokeAsync<IdbTransactionDurability>("durability");
            return result;
        }

        public async ValueTask<IWrappedDomException?> GetErrorAsync()
        {
            var result = await this.WrappedObject.InvokeObjectOrNullAsync<IJSObjectReference>("error");
            if (result is null)
                return null;
            return new WrappedDomException(result);
        }

        public async ValueTask<IdbTransactionMode> GetModeAsync()
        {
            var result = await this.WrappedObject.InvokeAsync<IdbTransactionMode>("mode");
            return result;
        }

        public async ValueTask<string[]> GetObjectStoreNamesAsync()
        {
            var result = await this.WrappedObject.InvokeAsync<string[]>("objectStoreNames");
            return result;
        }

        public async ValueTask<IWrappedIdbObjectStore> ObjectStoreAsync(string name)
        {
            var result = await this.WrappedObject.InvokeAsync<IJSObjectReference>("objectStore", name);
            return new WrappedIdbObjectStore(result);
        }

        public async ValueTask SetOnAbortAsync(EventObjectOfIdbTransaction? callbackObject)
        {
            await this.WrappedObject.InvokeVoidAsync("setOnAbort", callbackObject);
        }

        public async ValueTask SetOnCompleteAsync(EventObjectOfIdbTransaction? callbackObject)
        {
            await this.WrappedObject.InvokeVoidAsync("setOnComplete", callbackObject);
        }

        public async ValueTask SetOnErrorAsync(EventObjectOfIdbTransaction? callbackObject)
        {
            await this.WrappedObject.InvokeVoidAsync("setOnError", callbackObject);
        }
    }
}
